src/ls_server.rs

```diff
@@ -245,16 +252,35 @@ impl LsService {
         self.handler.init(init.rootPath, &*self.output);
     }
 
+    fn shutdown(&self, id: usize) {
+        self.pending_shut_down.store(true, Ordering::SeqCst);
+        self.output.success(id, serde_json::to_string(&serde_json::Value::Null).unwrap());
+    }
+    fn exit(&self, _id:usize) {
+        if !self.pending_shut_down.load(Ordering::SeqCst) {
+            // TODO Set exit code.
+            self.logger.log("Exit without shutting down first!");
+        }
+        self.shut_down.store(true, Ordering::SeqCst);
+        ::std::sync::atomic::fence(Ordering::SeqCst);
```

Comment:
I think that I need this fence, as closing stdin unblocks the other thread, which then immediately reads the atomic bool.  If I didn't have it, I think the compiler would be within it's rights to re-order the close call before the store, would it not?